pkg/metrics/metrics.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/apimachinery/pkg/types"
+	k8smetrics "k8s.io/component-base/metrics"
+	"k8s.io/klog"
+)
+
+const (
+	labelDriverName      = "driver_name"
+	labelOperationName   = "operation_name"
+	labelOperationStatus = "operation_status"
+	subSystem            = "snapshot_controller"
+	metricName           = "operation_total_seconds"
+	metricHelpMsg        = "Total number of seconds spent by the controller on an operation from end to end"
+)
+
+type OperationState string
+
+const (
+	// Failed is an operation state which means the controller has encountered
+	// an error which is not recoverable and the controller has marked a permanent
+	// failure of the operation.
+	Failure OperationState = "Failure"
```

Comment:
In what situations does the controller not retry?

Do we also count user cancellation as a failure?

Reply:
ATM when CSI failed permanently it would stop retrying to call CSI.
when you say user cancellation, are you referring to deletion of a VS?

Reply:
What is failed permanently? I don't think we should stop retrying until user cancels by deleting the VS. It is a Kubernetes anti-pattern

Reply:
This failure refers to the failure of one try.  The controller should always re-try.  We do have a bug to be fixed - If create snapshot fails very early, before the request is sent to the CSI driver, there will be no retries.  But we do plan to fix it when fixing the requeue logic.

Reply:
one concern is that the cache will grow linearly to the number of failed volumesnapshots over time, and when should we report an error metrics if there is no permanently failed VolumeSnapshot?

Reply:
the only time this e2e metric should be recorded is:
* volume operation is successful
* volume operation was aborted

Reply:
"volume operation was aborted" - create snapshot will only be aborted when user deletes the snapshot, but how is that a failure?

Reply:
A couple scenarios I see are:
* There was an invalid configuration so user canceled the operation, fixed the config, and then retry
* There was a system error and the user gave up and deleted the object
* Snapshot was taking too long and the user got impatient and gave up.

Reply:
If reconcile will retry CSI calls regardless the CSI return code (which is not the case ATM), I feel having an error metrics for end to end operation is not necessary and maybe not that valuable. User cancellation/abort is a pretty random behavior, they could cancel anytime after a failure has been observed, which will in turn make the metrics latency value not that useful. On the other hand, the error count, could be potentially valuable, as it would give users an idea how many total cancellations have been observed.

Reply:
The latency value could tell us information about how long users have waited before giving up. I think since count is included as part of the histogram anyway, it doesn't hurt to include the latency. Consumers can decide if they want the use the data or not.

Reply:
(Also we should fix the controller to retry in all cases. This was one of the design changes we made for beta)